transport/swift/swift.go

```diff
@@ -0,0 +1,107 @@
+package swift
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/imgproxy/imgproxy/v3/config"
+	"github.com/ncw/swift/v2"
+)
+
+type transport struct {
+	con *swift.Connection
+}
+
+func New() (http.RoundTripper, error) {
+	c := &swift.Connection{
+		UserName:       config.SwiftUsername,
+		ApiKey:         config.SwiftAPIKey,
+		AuthUrl:        config.SwiftAuthURL,
+		AuthVersion:    config.SwiftAuthVersion,
+		Domain:         config.SwiftDomain, // v3 auth only
+		Tenant:         config.SwiftTenant, // v2 auth only
+		Timeout:        time.Duration(config.SwiftTimeoutSeconds) * time.Second,
+		ConnectTimeout: time.Duration(config.SwiftConnectTimeoutSeconds) * time.Second,
+	}
+
+	ctx := context.Background()
+
+	err := c.Authenticate(ctx)
+
+	if err != nil {
+		return nil, fmt.Errorf("swift authentication error: %s", err)
+	}
+
+	return transport{con: c}, nil
+}
+
+func (t transport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	container, path, err := parseObjectURL(req.URL.Path)
+
+	if err != nil {
+		return &http.Response{
+			Status:     "400 Bad Request",
+			StatusCode: 400,
+			Proto:      "HTTP/1.1",
+			ProtoMajor: 1,
+			ProtoMinor: 1,
+			Close:      true,
+			Request:    req,
+		}, err
+	}
+
+	headers := make(swift.Headers)
+
+	object, headers, err := t.con.ObjectOpen(req.Context(), container, path, false, headers)
+
+	if err != nil {
+		return nil, fmt.Errorf("error opening object: %v", err)
+	}
+
+	header := make(http.Header)
+
+	if config.ETagEnabled {
+		if etag, ok := headers["Etag"]; ok {
+			header.Set("ETag", etag)
+
+			if len(etag) > 0 && etag == req.Header.Get("If-None-Match") {
```

Comment:
I believe we should close `object` here

Reply:
Yup, I checked, we MUST close `object` here

Reply:
yeah I added `defer object.Close()` at https://github.com/imgproxy/imgproxy/pull/837/files#diff-a187eae60042edf0861f826dce91206512effa21002f261eea57c5e19812a52fR55

Reply:
With `defer object.Close()` in that place, we close `object` no matter what. So if we respond with 200, the response body will be closed. We should close `object` only when we respond with 304 as we don't pass it with the response.

Reply:
You are right. I'm an idiot 😢 .  That's fixed now. Tested manually.

```
INFO    [2022-04-06T20:46:41+10:00] Completed in 545.22448ms /rs:fill/w:800/h:600/plain/swift://test/test.png  request_id=artKs1EMWHiGl3pncT0Fi method=GET status=200 client_ip="::1" image_url="swift://test/test.png" processing_options="Width: 800; Height: 600; Format: png"
INFO    [2022-04-06T20:46:42+10:00] Started /favicon.ico  request_id=eAj04jxtHh2BOtJqPjTFp method=GET client_ip="::1"
```

Reply:
Take it easy, could happen to everyone :)